File: packages/sdk-py/src/smscode/retry.py
from __future__ import annotations

import asyncio
import random
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import TypeVar

T = TypeVar("T")

BACKOFF_BASE_SECONDS = 0.250
BACKOFF_CAP_SECONDS = 10.0


@dataclass(frozen=True)
class RetryPolicy:
    max_retries: int = 0
    retry_on: Callable[[Exception], bool] = lambda err: False
    retry_after: Callable[[Exception], float | None] | None = None
    delay_seconds: Callable[[int], float] | None = None
    sleep: Callable[[float], None] | None = None
    async_sleep: Callable[[float], Awaitable[None]] | None = None


def default_delay_seconds(attempt: int) -> float:
    ceiling = min(BACKOFF_CAP_SECONDS, BACKOFF_BASE_SECONDS * 2 ** max(0, attempt - 1))
    return float(random.random() * ceiling)


def _retry_delay_seconds(err: Exception, attempt: int, policy: RetryPolicy) -> float:
    retry_after = policy.retry_after(err) if policy.retry_after is not None else None
    if retry_after is not None:
        return max(0.0, retry_after)
    delay = (
        policy.delay_seconds(attempt)
        if policy.delay_seconds is not None
        else default_delay_seconds(attempt)
    )
    return max(0.0, delay)


def with_retry(fn: Callable[[], T], policy: RetryPolicy) -> T:
    sleep = policy.sleep or time.sleep
    max_retries = max(0, policy.max_retries)
    for attempt in range(max_retries + 1):
        try:
            return fn()
        except Exception as err:
            if attempt >= max_retries or not policy.retry_on(err):
                raise
            sleep(_retry_delay_seconds(err, attempt + 1, policy))
    raise RuntimeError("unreachable retry loop")


async def async_with_retry(fn: Callable[[], Awaitable[T]], policy: RetryPolicy) -> T:
    sleep = policy.async_sleep or asyncio.sleep
    max_retries = max(0, policy.max_retries)
    for attempt in range(max_retries + 1):
        try:
            return await fn()
        except Exception as err:
            if attempt >= max_retries or not policy.retry_on(err):
                raise
            await sleep(_retry_delay_seconds(err, attempt + 1, policy))
    raise RuntimeError("unreachable retry loop")

File: packages/sdk-js/src/retry.ts
/**
 * Generic retry engine for transient failures.
 *
 * {@link withRetry} runs `fn`, and on a thrown error consults the {@link RetryPolicy}
 * to decide whether to retry, how long to wait, and stops after `maxRetries`
 * retries (re-throwing the last error). It is the **single** retry implementation
 * in the SDK: {@link SmscodeClient.request} delegates its transient-failure retry
 * to this function, and `orders.create` rides that same path, so there is exactly
 * one place that owns "wait and try again".
 *
 * The wait is `retryAfter(err)` (a server `Retry-After`, in **seconds**) when
 * present, otherwise capped exponential backoff with jitter. Both the wait clock
 * (`sleep`) and the backoff (`delayMs`) are injectable so tests stay instant and
 * deterministic.
 */

/** A policy that decides whether/how to retry a failed attempt. */
export interface RetryPolicy {
  /** Maximum number of *retries* (so total attempts = `maxRetries + 1`). */
  maxRetries: number;
  /** Whether a given thrown error is worth retrying. */
  retryOn: (err: unknown) => boolean;
  /**
   * Seconds to wait before the next retry, sourced from the error (e.g. a
   * `Retry-After` header). Returns `undefined` to fall back to backoff.
   */
  retryAfter?: (err: unknown) => number | undefined;
  /**
   * Backoff for retry attempt `n` (1-based) in milliseconds, used when
   * `retryAfter` yields nothing. Defaults to capped exponential backoff + jitter.
   */
  delayMs?: (attempt: number) => number;
  /** The wait clock. Defaults to a real `setTimeout`-backed sleep. Injectable for tests. */
  sleep?: (ms: number) => Promise<void>;
}

/** Base backoff step (ms). The first retry waits ~this, doubling each step. */
const BACKOFF_BASE_MS = 250;
/** Backoff ceiling (ms) so a long outage does not produce multi-minute waits. */
const BACKOFF_CAP_MS = 10_000;

/** Default `setTimeout`-backed sleep. */
function defaultSleep(ms: number): Promise<void> {
  if (ms <= 0) return Promise.resolve();
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Capped exponential backoff with full jitter for retry attempt `n` (1-based):
 * `random(0, min(cap, base * 2^(n-1)))`. Jitter spreads retries to avoid
 * thundering-herd reconnects; the result is still monotonic in expectation.
 */
function defaultDelayMs(attempt: number): number {
  const exp = Math.min(BACKOFF_CAP_MS, BACKOFF_BASE_MS * 2 ** (attempt - 1));
  // Full jitter keeps the upper bound growing while spreading load.
  return Math.random() * exp;
}

/**
 * Run `fn` with retry on transient failure.
 *
 * @typeParam T  The value `fn` resolves to.
 * @param fn      The operation to run (re-invoked on each retry).
 * @param policy  The retry policy.
 * @returns The first successful value.
 * @throws The last error once retries are exhausted or `retryOn` returns `false`.
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  policy: RetryPolicy,
): Promise<T> {
  const sleep = policy.sleep ?? defaultSleep;
  const delayMs = policy.delayMs ?? defaultDelayMs;
  const maxRetries = Math.max(0, policy.maxRetries);

  let lastError: unknown;
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastError = err;
      const hasRetryBudget = attempt < maxRetries;
      if (!hasRetryBudget || !policy.retryOn(err)) {
        throw err;
      }
      const retryAfterSecs = policy.retryAfter?.(err);
      const waitMs =
        retryAfterSecs !== undefined && Number.isFinite(retryAfterSecs)
          ? Math.max(0, retryAfterSecs) * 1000
          : delayMs(attempt + 1);
      await sleep(waitMs);
    }
  }
  // Unreachable: the loop returns on success or throws on the final attempt.
  throw lastError;
}
